### DISCUSSION TITLE
Welcome to vue3-realworld-example-app Discussions!

### DISCUSSION THREAD

**Discussion by mutoe** [mutoe/vue3-realworld-example-app#47] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by ashish-banjare** [comment]
Hello there,

thanks for your contribution 
i am new in Vuejs, for learning purpose i setuped this repo with laravel realworld backend
i am getting this error while update profile or create new article.
**{"errors":{"message":"JWT error: Token is absent","status_code":401}}**

**Reply by mutoe** [reply]
You should set `Authorization` header in your request.

```
Authorization: Token xxx
```

> https://en.wikipedia.org/wiki/JSON_Web_Token

**Comment by dmshved** [comment]
Hello Mr. mutoe

After analyzing the source code I still have some questions:

- Why is userStorage (localStorage) used to store Token if it makes the token vulnerable to XSS attacks?
- What is the point of moving the token to securityData in JS memory if it is still retrieved from localStorage beforehand? 
- Is there any additional protection provided, or is this done to simplify the implementation of a RealWorld project?

I'd be grateful for your expert opinion, I just want to understand how I can structure my VueJS app and work with backend authorization

**Reply by mutoe** [reply]
Hello Dmytro! 

First of all, I am glad that you are paying attention to this project. Regarding your questions, I also have some questions that I can't answer, because it seems that this is what everyone has done since I started working. 
Therefore, I specifically asked AI. There is no problem with your concerns, but considering the security requirements and safety requirements of the current project, it really does not need to be so complicated.
 Security and convenience have been two aspects that cannot be achieved at the same time since ancient times. For example, generally high-security software requires you to set a password of more than 8 characters, which must also contain numbers, uppercase and lowercase letters, and special symbols, but I still often forget the password. The password is too simple, which reduces security, and it is too complex and inconvenient for me to use. So we need to balance them appropriately.

AI's answer is as follows:

---

## 💡 Expert Analysis: JWT Storage and Security in RealWorld VueJS

The RealWorld project (including the VueJS implementation) is designed as a **full-featured example** to demonstrate how a frontend framework interacts with a common API specification (like the Conduits API). Its primary goal is to showcase **functional implementation and architectural patterns**, rather than offering an **ultimate production-grade** security solution.

### 1. Why is `localStorage` used to store the Token if it makes it vulnerable to XSS attacks?

* **RealWorld's Intent (Simplicity and Demonstration):**
    * In many tutorials and basic examples, using `localStorage` to store the JSON Web Token (JWT) is the **simplest and most accessible** method. It allows the Token to persist across page refreshes and can be easily accessed by JavaScript to be attached to subsequent API request headers (e.g., `Authorization: Token <jwt>`).
    * For a **demonstration project** like RealWorld, this simplification is generally **acceptable** because it allows developers to focus on the **core framework logic** of Vue/Vuex/Routing.
* **Security Risk (XSS Vulnerability):**
    * You are absolutely correct! Storing the Token in `localStorage` or `sessionStorage` makes it **highly vulnerable to Cross-Site Scripting (XSS) attacks**.
    * If an attacker can inject a malicious script onto your site (e.g., via a comments section or profile field), that script can easily use `localStorage.getItem('jwt')` to steal the user's Token and impersonate the user by sending API requests.

* **Production Environment Best Practice (Recommendation):**
    * In a **production environment**, the standard recommendation is to store the Token (or a Session ID) in an **HttpOnly Cookie**.
    * **Advantages of HttpOnly Cookie:**
        * A Cookie set as `HttpOnly` **cannot** be accessed by client-side JavaScript (like `document.cookie` or the `localStorage` API), which significantly mitigates the risk of **XSS attacks** stealing the Token.
        * The Token is **automatically** attached by the browser to every relevant request, simplifying client-side code.
        * **Note:** HttpOnly Cookies are still vulnerable to **CSRF (Cross-Site Request Forgery)** attacks. Therefore, you must combine them with the **SameSite Cookie attribute (recommended `SameSite=Strict` or `Lax`)** and/or a **CSRF Token mechanism** for comprehensive protection.
        

---

### 2. What is the point of moving the token to securityData in JS memory if it is still retrieved from localStorage beforehand?

The primary purpose of reading the Token from `localStorage` and then storing it in a Vuex Store or another JS memory variable (like `securityData`) is **not to enhance security**, but rather to **facilitate application state management and access**:

* **Core State Management Need:**
    * In a Single Page Application (SPA), you need a **centralized, reactive** place to store application-level state like "is the user logged in?" and "what is the user's Token?". Vuex (or Pinia/other state management tools) is designed for this.
    * Placing the Token in JS memory (e.g., in Vuex `state`) ensures that **any component or service across the application** can quickly and consistently access it for:
        1.  Checking authentication status in route guards.
        2.  Easily attaching it to the `Authorization` header of every outgoing API request within the API service layer.
* **Not a Security Mechanism:**
    * As you rightly observed, because the Token is **initially retrieved from the insecure `localStorage`**, moving it to JS memory **adds no extra security protection** against an XSS attack. A malicious script that can read `localStorage` can also easily read the Vuex Store's `securityData`.

---

### 3. Is there any additional protection provided, or is this done to simplify the implementation of a RealWorld project?

* **Conclusion:** This is primarily done to **simplify implementation and demonstrate common SPA architectural patterns**, and **it does not provide additional security protection** against the most critical client-side threat (XSS).

* **Summary and Architectural Guidance:**

| Goal | RealWorld VueJS Example Approach | Production-Ready Recommended Approach |
| :--- | :--- | :--- |
| **Persistence Storage** | `localStorage` | **HttpOnly Cookie** (with SameSite attribute) |
| **Token Accessibility** | Accessible by JS scripts (High XSS risk) | Inaccessible by JS scripts (Mitigates XSS) |
| **State Management** | Loads Token from `localStorage` into Vuex/Pinia state | Authentication status (e.g., `isAuthenticated: true/false`) can be derived from the backend response or cookie status (Token remains in the Cookie). |
| **Design Intention** | Simplify implementation, showcase Vue/Vuex/Router architecture | Ensure functional correctness with maximum security posture. |

**My Advice for Your App:**

If you are building an application that handles real user data, **do not** use the RealWorld Token storage approach for your production environment.

The secure approach involves:

1.  **Upon successful login**, the backend sends the JWT (or Session ID) in an **`HttpOnly`**, **`Secure`**, and **`SameSite=Lax/Strict`** **Cookie** back to the client.
2.  The **frontend code** does **not** directly touch this Token. The browser automatically sends it with requests to the backend.
3.  The **frontend** only needs to maintain a simple boolean in memory (e.g., `isAuthenticated: true/false`) to represent the user's logged-in status, rather than storing the Token itself.

I hope this detailed explanation helps you understand the trade-offs in the RealWorld project and guides you towards a more secure design for your own VueJS application!

**Reply by dmshved** [reply]
Thanks for your apply, actually I was thinking about this way so code would be more convenient and easy to understand. After some researches and discussions with other developers I came up with the Session Based Authentication, since I am currently working with Laravel I can use Laravel Sanctum with his built-in session based auth using session and csrf tokens. 

Im leaving a link here for someone who will find this discussion: https://github.com/Dmytro-Shved/welltrip-vue

Very grateful for your reply and your project, it gave me some valuable practices I couldnlt find before

Regards,

Dmytro Shved
